### PR TITLE
Update Network Graph to use URL parameters for search.

### PR DESCRIPTION
**Pull request by dvail**
## Description

This updates the Network Graph page to use URL parameters for search filters to match the functionality and conventions used on other pages. This does not store other Network Graph state in the URL; in order to keep the scope of changes small this has been extracted from the larger WIP PR https://github.com/stackrox/stackrox/pull/1532.

## Checklist
- [ ] Investigated and inspected CI test results

## Testing Performed

Test that applying a search filter correctly filters the graph display and tracks search filters in the URL.
image

Test that applying multiple search filters correctly filters the graphs and tracks entries in the URL:
image

Test that the back button correctly removes one filter at a time:
image

Test that global search correctly applies the search filters to the graph and URL when visiting via the "Filter On" link:
(Note that the selected cluster and namespaces _will not_ change with these searches, and functions identically to `master`. Work to get context aware switching of clusters/namespaces is part of https://github.com/stackrox/stackrox/pull/1532) 
image
image

Test that clicking a namespace or deployment within the graph correctly updates the URL and page state without clearing search parameters.
image

Test that opening the CIDR block side panel does not clear search results:
image

Test the opening the Net Policy Simulator does not clear search results:
image

### CODE REVIEW

**Comment by pedrottimark** [comment]
Here is an existing issue, found while testing, which might be outside the scope of this pull request.

If `onSearch` callback in `NetworkSearch` component closes search panel because search filter might exclude the currently selected deployment, then it seems more consistent to remove deployment id from page url instead.

Both master and branch have the following behavior:

1. Visit **Network Graph**
2. Select Namespaces: **stackrox**
3. Add **Deployment:** and **scanner-db** to search filter
4. Click **scanner** deployment in graph: side panel opens and url now has deployment id
5. Delete **scanner-db** from search filter
6. Add **collector** to search filter: side panel closes **but url still has deployment id**

    Branch has the following additional behavior:

7. Go directly to original **Network Graph** in browser history: side panel is closed and search filter is empty (expected)
8. Go directly to last **Network Graph** in browser history: side panel is open for **scanner** because deployment id and search filter has **Deployment:** **collector** (expected according to code but inconsistent with step 6)

File: ui/apps/platform/src/utils/searchUtils.ts [hunk 1]
```diff
@@ -107,6 +107,16 @@ export function searchOptionsToSearchFilter(searchOptions: SearchEntry[]): Searc
...
 
+/**
+ * Determines whether or not a SearchFilter contains a valid value for
+ * all keys. A valid value is either a non-empty string or non-empty array.
+ */
+export function isCompleteSearchFilter(searchFilter: SearchFilter) {
+    return Object.values(searchFilter).every(
+        (o) => Boolean(o) && (!Array.isArray(o) || o.length > 0)
+    );
+}
```
Comment (vjwilson) [inline ui/apps/platform/src/utils/searchUtils.ts:118]: I think there is similar functionality in the WorkflowState class. 
(Unfortunately, that is written in classic OO style, and has become something of a "god-class", so I'm not sure it's worth trying to reconcile what's in it with helper functions like this one.)

**Review by vjwilson — APPROVED** [review]
/lgtm

Just a couple of observations

**Comment by dvail** [comment]
> If onSearch callback in NetworkSearch component closes search panel because search filter might exclude the currently selected deployment, then it seems more consistent to remove deployment id from page url instead.

I agree, and I found that as I was working on #1532 that adding URL parameters to track state and the current URL structure could be a little confusing. With the deployment id as part of the URL, it seems the full URL path (without query string) should point to a definitive resource within the graph and open side panels, etc. that are appropriate. Right now from what I can tell, only a selected deployment is supported with a structure of
`https://staging.demo.stackrox.com/main/network/<deployment_id>`

Since users can also select a full namespace or external entities it might make sense in the future to migrate the base URL structure to something like:
`https://staging.demo.stackrox.com/main/network/deployment/<deployment_id>`
`https://staging.demo.stackrox.com/main/network/namespace/<namespace_id>`
`https://staging.demo.stackrox.com/main/network/external/<external_entity>`
...and possibly more if there are other use cases I missed. That way we could track the "selected" state and render appropriately along side whatever query string parameters are applied.

For this:
> Go directly to last Network Graph in browser history: side panel is open for scanner because deployment id and search filter has Deployment: collector (expected according to code but inconsistent with step 6)

how exactly did you reproduce this issue?

If I go directly to the following URL on my branch: https://localhost:3000/main/network/966f30dd-a451-45bf-9a7e-7a5f505db7f0?s[Deployment]=collector, I get the following view:
image

If I go to the first Network Graph page in my browser's history though, and step through with the forward button to get to the above URL, I get the behavior you describe:
image

This makes sense (somewhat) because the side panel open state isn't really a declarative, derived state. From what I can tell, other state changes may trigger an action that sets the open/closed state on the panel but the panel state and the rest of the page state can get out of sync.

A possible enhancement for this in the future could be to have the panel state tied to a URL scheme like I suggested above, since it seems that is what the intended behavior is. Additionally we could easily tie in the selected panel tab with something like:
`https://staging.demo.stackrox.com/main/network/deployment/<deployment_id>/flows`
`https://staging.demo.stackrox.com/main/network/deployment/<deployment_id>/network-policies`
`https://staging.demo.stackrox.com/main/network/deployment/<deployment_id>/details`

Your suggested fix of clearing the selected deployment from the URL `onSearch` would be a good solution to the first issue, but I'm not sure that it will resolve the second.

File: ui/apps/platform/src/Containers/Network/Header/NetworkSearch.js [hunk 2]
```diff
@@ -1,29 +1,69 @@
...
+    Object.entries(searchFilter).forEach(([key, value]) => {
+        entries.push({ label: `${key}:`, value: `${key}:`, type: 'categoryOption' });
+        if (value !== '') {
+            const values = Array.isArray(value) ? value : [value];
+            const valueOptions = values.map((v) => ({ label: v, value: v }));
+            entries.push(...valueOptions);
+        }
+    });
+    return entries;
+}
```
[the remark this answers is not in the document]
Comment (dvail) [reply ui/apps/platform/src/Containers/Network/Header/NetworkSearch.js:31]: Hmm good call, and each one has a very slight variation in functionality that prevents a simple one-for-one replacement. I'll make a note to follow up and clean these up as an independent change.

File: ui/apps/platform/src/utils/searchUtils.ts (hunk 1 above)
Comment (pedrottimark) [reply ui/apps/platform/src/utils/searchUtils.ts:118]: Yeah, maybe whenever attention turns back to workflow, we might factor out some workflow methods to become available to call as pure functions.

**Comment by pedrottimark** [comment]
> Since users can also select a full namespace or external entities it might make sense in the future to migrate the base URL structure to something like

Yes, possible smaller and larger steps toward the goal for page url to become source of truth:

1. Include namespace in search query params. It seems worthwhile to explore cost and benefit of usual format `s[Namespace]=namespaceId` to hide implementation detail whether interaction is via combined or separate widgets (which might even change in the future).

2. Use nouns in url to represent selected item or interaction panel:
    * /main/network/deployments/deploymentId
    * /main/network/namespaces/namespaceId
    * /main/network/entities/TODO needs investigation what is the "id" for external entities
    * /main/network/blocks
    * /main/network/simulator
    * example of possible future: /main/network/policies/policyId

What are pro and con of singular versus plural nouns, which follows example of usual REST API endpoints?

Do any future concepts have **tables** of deployments, namespaces, or entities, corresponding to url without id?

**Review by pedrottimark — APPROVED** [review]
Super step forward. Keep up the good work.

**Comment by dvail** [comment]
@vjwilson @pedrottimark , for awareness since you already approved:

I found another issue after some exploration based on Mark's comments above that clears search parameters when the user clicks the CIDR Blocks or Net Policy buttons. Since those buttons push to the history stack we need to manually retain the `search` part of the URL.

The last commit I pushed addresses this and I added two screenshots to the test section on the PR.

File: ui/apps/platform/src/Containers/Network/SidePanel/useNavigateToEntity.ts [hunk 3]
```diff
@@ -9,9 +9,9 @@ function useNavigateToEntity(): NavigateToEntityHook {
     const history = useHistory();
     return function onNavigateToEntityById(entityId: string, type: EntityType): void {
         if (type === nodeTypes.CIDR_BLOCK || type === nodeTypes.EXTERNAL_ENTITIES) {
-            history.push(`/main/network/${entityId}/${type}`);
+            history.push(`/main/network/${entityId}/${type}${history.location.search as string}`);
```
Comment (pedrottimark) [inline ui/apps/platform/src/Containers/Network/SidePanel/useNavigateToEntity.ts:12]: Interesting investigation for the future whether this type of url is possible (pardon pun).

**Pull request merged**